libcaf_test/caf/test/fixture/deterministic.test.cpp

```diff
@@ -0,0 +1,346 @@
+
+#include "caf/test/fixture/deterministic.hpp"
+
+#include "caf/test/caf_test_main.hpp"
+#include "caf/test/test.hpp"
+
+#include "caf/event_based_actor.hpp"
+#include "caf/scoped_actor.hpp"
+#include "caf/send.hpp"
+
+namespace fixture = caf::test::fixture;
+
+struct my_int {
+  int value;
+};
+
+template <class Inspector>
+auto inspect(Inspector& f, my_int& x) {
+  return f.object(x).fields(f.field("value", x.value));
+}
+
+bool operator==(my_int lhs, my_int rhs) noexcept {
+  return lhs.value == rhs.value;
+}
+
+bool operator==(my_int lhs, int rhs) noexcept {
+  return lhs.value == rhs;
+}
+
+bool operator==(int lhs, my_int rhs) noexcept {
+  return lhs == rhs.value;
+}
+
+bool operator!=(my_int lhs, my_int rhs) noexcept {
+  return lhs.value != rhs.value;
+}
+
+CAF_BEGIN_TYPE_ID_BLOCK(test, caf::first_custom_type_id)
+
+  CAF_ADD_TYPE_ID(test, (my_int))
+
+CAF_END_TYPE_ID_BLOCK(test)
+
+WITH_FIXTURE(fixture::deterministic) {
+
+TEST("the deterministic fixture provides a deterministic scheduler") {
+  auto initialized = std::make_shared<bool>(false);
+  auto count = std::make_shared<int32_t>(0);
+  auto worker = sys.spawn([initialized, count] {
+    *initialized = true;
+    return caf::behavior{
+      [count](int32_t value) { *count += value; },
+      [count](const std::string& str) {
+        if (auto ival = caf::get_as<int32_t>(caf::config_value{str}))
+          *count += *ival;
+      },
+    };
+  });
+  caf::scoped_actor self{sys};
+  check(*initialized);
+  check_eq(mail_count(worker), 0u);
+  anon_send(worker, 1);
+  check_eq(mail_count(worker), 1u);
+  self->send(worker, 2);
+  check_eq(mail_count(worker), 2u);
+  anon_send(worker, 3);
+  check_eq(mail_count(worker), 3u);
+  SECTION("calling dispatch_message processes a single message") {
+    check(dispatch_message());
+    check_eq(mail_count(worker), 2u);
+    check_eq(*count, 1);
+    check(dispatch_message());
+    check_eq(mail_count(worker), 1u);
+    check_eq(*count, 3);
+    check(dispatch_message());
+    check_eq(mail_count(worker), 0u);
+    check_eq(*count, 6);
+    check(!dispatch_message());
+  }
+  SECTION("calling dispatch_messages processes all messages") {
+    check_eq(dispatch_messages(), 3u);
+    check_eq(*count, 6);
+  }
+#ifdef CAF_ENABLE_EXCEPTIONS
+  SECTION("expect() checks for required messages") {
+    expect<int32_t>().to(worker);
+    check_eq(mail_count(worker), 2u);
+    check_eq(*count, 1);
+    expect<int32_t>().to(worker);
+    check_eq(mail_count(worker), 1u);
+    check_eq(*count, 3);
+    expect<int32_t>().to(worker);
+    check_eq(mail_count(worker), 0u);
+    check_eq(*count, 6);
+    should_fail_with_exception([this, worker] { //
+      expect<int32_t>().with(3).to(worker);
+    });
+  }
+  SECTION("expect() matches the types of the next message") {
+    anon_send(worker, "4");
+    should_fail_with_exception([this, worker] { //
+      expect<std::string>().to(worker);
+    });
```

Comment:
Just to make sure I got this: `should_fail_with_exception` is for the purpose of testing the testing framework. In regular unit tests it should not be use and expect behaves like a `require` - it either succeeds or aborts the test.